rfcs/2020-08-27-3603-postgres-metrics.md

```diff
@@ -0,0 +1,157 @@
+# RFC 3603 - 2020-08-27 - Collecting metrics from PostgreSQL
+
+This RFC is to introduce a new metrics source to consume metrics from PostgreSQL database servers. The high level plan is to implement one source that collects metrics from PostgreSQL server instances.
+
+Background reading on PostgreSQL monitoring:
+
+- https://www.datadoghq.com/blog/postgresql-monitoring/
+
+## Scope
+
+This RFC will cover:
+
+- A new source for PostgreSQL server metrics.
+
+This RFC will not cover:
+
+- Other databases.
+
+## Motivation
+
+Users want to collect, transform, and forward metrics to better observe how their PostgreSQL databases are performing.
+
+## Internal Proposal
+
+Build a single source called `postgresql_metrics` (name to be confirmed) to collect PostgreSQL metrics. We support all non-EOL'ed PostgreSQL versions.
+
+The recommended implementation is to use the Rust PostgreSQL client to connect the target database server by address specified in configuration.
+
+- https://docs.rs/postgres/0.17.5/postgres/index.html
+
+
+The source would then run the following queries:
+
+- `SELECT * FROM pg_stat_database`
+- `SELECT * FROM pg_stat_database_conflicts`
+- `SELECT * FROM pg_stat_bgwriter`
+
+And return these metrics by parsing the query results and converting them into metrics using the database name and column names.
+
+- `pg_up` -> Used as an uptime metric with 0 for successful collection and 1 for failed collection (gauge)
```

Comment:
Do you expect these all to be prefixed `pg_` along with the default namespace of `postgresql`? That would make, for example, `pg_stat_database_blks_hit_total` publish as `postgresql_pg_stat_database_blks_hit_total` to prometheus.

Reply:
I am not 100% clear on namespaces and their role?

Reply:
I think their role is to:

* Segment metrics that might otherwise collide and avoid aggregations across them
* Group metrics together logically

For prometheus and statsd I think they just manifest as a prefix, but other metrics sinks, like [`CloudWatch`](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_concepts.html#Namespace) they are a separate field.

Related #3609 . I think we may need to make metric namespace a first-class field in vector to be able to send it separately to sinks like `cloudwatch_metrics` that require it as a separate field rather than a prefix.

Reply:
Prometheus also calls it out in https://prometheus.io/docs/practices/naming/ (referring to it as "domain").

Reply:
Given that we're adopting Prometheus' rules, it should be simple enough to bake this into the metric name, right?

> a (single-word) application prefix

And I agree, it seems that we'd want to require this so it's always present.

Reply:
Perhaps we should throw together an RFC on metrics design?

Reply:
Yeah. #3600 represents the task. How about we add the rules to `CONRIBUTING.md`, as described in that issue, and obtain consensus in that PR?

Reply:
Sounds good. We'll see how long they are perhaps and work out if we should have a page in the docs if they are a standalone size?